README
NAME

    Mojo::Base::XS - very fast Mojo-styled accessors

INSTALLATION

        perl Makefile.PL
        make
        make test
        make install

SYNOPSIS

        # Replace Mojo::Base
        use Mojo::Base::XS -infect;
        package Foo;
        use Mojo::Base;

        has foo => 'bar';
        has [qw/x y z/] => 42;
        has foo_defaults => sub { Bar->new };

        # Or use as standalone accessor generator
        package Bar;
        use Mojo::Base::XS;

        has bar => 'bar';
        has [qw/x y z/] => 42;
        has bar_defaults => sub { Baz->new };

    Mojo::Base::XS also allows you to run existing applications
    without any modifications:

        perl -MMojo::Base::XS=-infect script/your_app

DESCRIPTION

    Mojo::Base::XS implements fast accessrors for Mojo-based software.
    Code based on Class::XSAccessor - fastes Perl accessors.

    It can also be used as standalone Mojo-style accessors generator.

SEE ALSO

    * Class::XSAccessor
    * Mojo::Base

AUTHOR

    Yaroslav Korshak <yko@cpan.org>

LICENCE AND COPYRIGHT

    Copyright (C) 2011-2013, Yaroslav Korshak

    This module is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself. See perlartistic.


Changes
Revision history for Mojo-Base-XS

0.09  Sat Aug 03 20:43:17 2013
      Apply workaround for perl #117947 on affected versions
      Portad TARG-optimization from Class::XSAccessor
      Fixed couple lvalue-related issues
      Fixed #2 - broken perl < 5.14.0 compatibility

0.08  Wed May 08 22:58:42 2013
      Rebased to Class-XSAccessor-1.16 C code

0.07  Sun May 05 01:30:01 2013
      Added exceptions to match Mojo::Base behavior
      Deny calls to accessors on a classname

0.06  Mon Oct 22 14:19:28 2012
      Updated to Module::Install 1.06
      Separated author tests

0.05  Thu Nov 10 19:10:42 2011
      Fixed bug with ExtUtils::ParseXS > 3.04_02 and XSUBs export

0.04  Sun Aug 28 09:46:20 2011
      Require perl 5.10
      Made tests more failsafe

0.03  Sat Aug 27 13:38:02 2011
      Fixed examples

0.02  Sat Aug 27 13:23:12 2011
      Fixed memory leaks
      Fixed bug with generated default value
      Improved documentation

0.01  Fri Aug 12 13:40:56 2011
      Initial release.


t/lib/BaseTestXS.pm
package BaseTestXS;

use strict;
use warnings;

use Mojo::Base::XS;

has [qw/ears eyes/] => sub {2};

__PACKAGE__->attr(heads => 1);
__PACKAGE__->attr('name' => sub { 'Named!' });
__PACKAGE__->attr('def_array' => sub { ['Named!'] });
__PACKAGE__->attr('dies_in_default' => sub { die "Exception thrown" });

1;


t/lib/BaseTestXS/Base1.pm
package BaseTestXS::Base1;
use Mojo::Base::XS;

# "Okay folks, show's over. Nothing to see here, show's... Oh my god!
#  A horrible plane crash!
#  Hey everybody, get a load of this flaming wreckage!
#  Come on, crowd around, crowd around!"
has 'bananas';

1;


t/lib/BaseTestXS/Base2.pm
package BaseTestXS::Base2;
use Mojo::BaseXS;
use base 'BaseTestXS::Base1';

# "Hey, I asked for ketchup! I'm eatin' salad here!"
has [qw/ears eyes/] => sub {2};
has coconuts => 0;

1;


cxsa_main.h
#ifndef cxsa_main_h_
#define cxsa_main_h_

#include "EXTERN.h"
#include "perl.h"
#include "ppport.h"

#include "cxsa_hash_table.h"

typedef struct autoxs_hashkey_str autoxs_hashkey;
struct autoxs_hashkey_str {
  U32 hash;
  char* key;
  I32 len; /* not STRLEN for perl internal UTF hacks and hv_common_keylen
              -- man, these things can take you by surprise */
  autoxs_hashkey * next; /* Alas, this is the node of a linked list */

  /* It may be tempting to add more data here for further parameterization
   * of the accessor methods. That's simple at first sight (just add another
   * bit of data here, say a void* for user data, whatever). But when you
   * think about it some more, you'll see that this has various nasty
   * consequences in the context of re-using the hashkey entries.
   * - Hashkey structs are shared and there's a reverse look-up using the
   *   hash-key-string itself for that purpose.
   * - Sharing the hashkey structs is important for threads: Otherwise, we
   *   have to implement full cloning on thread creation. Oh my.
   * - If we didn't share hashkey structs, we'd have to deallocate them when
   *   an accessor method is no longer used. That means detecting when that
   *   happens. Note that there is not reference counting mechanism we can
   *   use to make this easy. So the sharing/reuse of hashkey structs was a
   *   way to avoid this real or perceived memory leak on re-generation of
   *   accessors for the same object slot.
   */
/* Mojo::Base::XS injection */
  char* accessor_name;
  bool  default_coderef;
  I32 accessor_len;
  SV *default_value;
/* END Mojo::Base::XS injection */

};

autoxs_hashkey * get_hashkey(pTHX_ const char* key, const I32 len);
I32 get_internal_array_index(I32 object_ary_idx);

/* Macro to encapsulate fetching of a hashkey struct pointer for the actual
 * XSUBs */
#define CXAH_GET_HASHKEY ((autoxs_hashkey *) XSANY.any_ptr)


/*************************
 * thread-shared memory
 ************************/

extern autoxs_hashkey * CXSAccessor_last_hashkey;
extern autoxs_hashkey * CXSAccessor_hashkeys;
extern HashTable* CXSAccessor_reverse_hashkeys;

extern U32 CXSAccessor_no_arrayindices;
extern U32 CXSAccessor_free_arrayindices_no;
extern I32* CXSAccessor_arrayindices;

extern U32 CXSAccessor_reverse_arrayindices_length;
extern I32* CXSAccessor_reverse_arrayindices;

#endif


lib/Mojo/Base/XS.pm
package Mojo::Base::XS;

use strict;
use warnings;

use 5.010;
require feature;

our $VERSION = 0.09;
require XSLoader;

XSLoader::load('Mojo::Base::XS', $VERSION);

my $base_infected;

sub import {
    my $class = shift;
    my $flag = shift || '';

    if ($flag eq '-infect') {
        return if $base_infected++;

        # Check if no Mojo::Base loaded
        if (UNIVERSAL::can('Mojo::Base', 'new')) {
            require Carp;
            Carp::croak(
                "You must load Mojo::Base::XS before loading Mojo::Base");
        }

        # Load and monkey-patch Mojo::Base
        require Mojo::Base;

        no strict 'refs';
        no warnings 'redefine';

        newxs_attr('Mojo::Base::attr');
        newxs_constructor('Mojo::Base::new');
    }
    else {
        my $caller = caller;

        no strict 'refs';
        no warnings 'redefine';

        newxs_attr($caller . '::attr');
        newxs_constructor($caller . '::new');

        # TODO: turn this into XS code
        *{"${caller}::has"} = sub { attr($caller, @_) };
    }
}

1;

__END__

=head1 NAME

Mojo::Base::XS - fast Mojo-styled accessors


=head1 SYNOPSIS

    # Replace Mojo::Base
    use Mojo::Base::XS -infect;
    package Foo;
    use Mojo::Base;

    has foo => 'bar';
    has [qw/x y z/] => 42;
    has foo_defaults => sub { Bar->new };

    # Or use as standalone accessor generator
    package Bar;
    use Mojo::Base::XS;

    has bar => 'bar';
    has [qw/x y z/] => 42;
    has bar_defaults => sub { Baz->new };

L<Mojo::Base::XS> also allows you to run existing applications
without any modifications:

    perl -MMojo::Base::XS=-infect script/your_app

=head1 DESCRIPTION

Mojo::Base::XS implements fast accessors for Mojo-based software.
This library is based on the code of L<Class::XSAccessor>.

It can also be used as standalone Mojo-style accessors generator.

=head1 EXPORTS

L<Mojo::Base::XS> exports following functions:

=head2 has

    has 'name';
    has [qw/name1 name2 name3/];
    has name => 'foo';
    has name => sub {...};
    has [qw/name1 name2 name3/] => 'foo';
    has [qw/name1 name2 name3/] => sub {...};

Create attributes for hash-based objects, just like the L<attr> method.

=head1 FUNCTIONS

L<Mojo::Base::XS> implements following functions:

=head2 newxs_attr

    newxs_attr('Class', 'attr');
    Class->attr(foo => 'bar');
    print Class->foo; # bar

Installs XS attribute generator subroutine

=head2 newxs_constructor

    newxs_attr('Class', 'new');
    Class->new(foo => 'bar');
    print Class->{foo}; # bar

Installs XS constructor

=head2 accessor, accessor_init, constructor, constructor_init

Inherited from L<Class::XSAccessors> for internal use

=head1 METHODS

L<Mojo::Base::XS> implements following methods:

=head2 C<new>

    my $obj = Class->new;
    my $obj = Class->new(name => 'value');
    my $obj = Class->new({name => 'value'});

This base class provides a basic object constructor.
You can pass it either a hash or a hash reference with attribute values.

=head2 C<attr>

    __PACKAGE__->attr('name');
    __PACKAGE__->attr([qw/name1 name2 name3/]);
    __PACKAGE__->attr(name                    => 'foo');
    __PACKAGE__->attr(name                    => sub {...});
    __PACKAGE__->attr([qw/name1 name2 name3/] => 'foo');
    __PACKAGE__->attr([qw/name1 name2 name3/] => sub {...});

Create attribute accessor for hash-based objects.
An arrayref can be used to create more than one attribute.
Pass an optional second argument to set a default value, it should be a
constant or a sub reference.
The sub reference will be executed at accessor read time if there's no set
value.
Accessors can be chained, that means they return their invocant
when they are called with an argument.

=head1 BUGS

L<Mojo::Base::XS>'s goal is to exactly match the behaviour
of all L<SUPPORTED VERSIONS OF Mojo::Base>.

Any differences in behavior between L<Mojo::Base::XS> and L<Mojo::Base>
should be considered as as a bug of L<Mojo::Base::XS>,
if it's not documented in L<DEFFERENCES WITH Mojo::Base>.

=head1 SUPPORTED VERSIONS OF Mojo::Base

TBD

=head1 DEFFERENCES WITH Mojo::Base

L<Mojo::Base::XS> has the following differences with L<Mojo::Base>,
which are intentional and should not be considered as bugs.

=over

=item * An accessor method call on a class name

    perl -MMojo::Base=-base -E 'has "bar" => 42; say main->bar'

=over

=item L<Mojo::Base>

Output:

    42

=item L<Mojo::Base::XS>

Throws the following exception:

    Accessor 'bar' should be called on an object, but called on the 'main' clasname

=back

=back

=head1 SEE ALSO

=over

=item * L<Class::XSAccessor>

L<Mojo::Base::XS> based on L<Class::XSAccessor> with modifications.

=item * L<Mojo::Base>

=back

=head1 AUTHOR

Yaroslav Korshak E<lt>yko@cpan.orgE<gt>

=head1 LICENCE AND COPYRIGHT

Copyright (C) 2011-2013, Yaroslav Korshak

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See L<perlartistic>.
